external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/InternalKafkaConsumerPool.scala

```diff
@@ -0,0 +1,246 @@
+  }
+
+  private def extractCacheKey(consumer: InternalKafkaConsumer): CacheKey = {
+    new CacheKey(consumer.topicPartition, consumer.kafkaParams)
+  }
+}
+
+private[kafka010] object InternalKafkaConsumerPool {
+
+  /**
+   * Builds the pool for [[InternalKafkaConsumer]]. The pool instance is created per each call.
+   */
+  def build: InternalKafkaConsumerPool = {
+    val objFactory = new ObjectFactory
+    val poolConfig = new PoolConfig
+    new InternalKafkaConsumerPool(objFactory, poolConfig)
+  }
+
+  object CustomSwallowedExceptionListener extends SwallowedExceptionListener with Logging {
+    override def onSwallowException(e: Exception): Unit = {
+      logError(s"Error closing Kafka consumer", e)
+    }
+  }
+
+  class PoolConfig extends GenericKeyedObjectPoolConfig[InternalKafkaConsumer] {
+    private var softMaxTotal = Int.MaxValue
+
+    def getSoftMaxTotal(): Int = softMaxTotal
+
+    init()
+
+    def init(): Unit = {
+      import PoolConfig._
+
+      val conf = SparkEnv.get.conf
+
+      softMaxTotal = conf.getInt(CONFIG_NAME_CAPACITY, DEFAULT_VALUE_CAPACITY)
+
+      val jmxEnabled = conf.getBoolean(CONFIG_NAME_JMX_ENABLED,
+        defaultValue = DEFAULT_VALUE_JMX_ENABLED)
+      val minEvictableIdleTimeMillis = conf.getLong(CONFIG_NAME_MIN_EVICTABLE_IDLE_TIME_MILLIS,
+        DEFAULT_VALUE_MIN_EVICTABLE_IDLE_TIME_MILLIS)
+      val evictorThreadRunIntervalMillis = conf.getLong(
+        CONFIG_NAME_EVICTOR_THREAD_RUN_INTERVAL_MILLIS,
+        DEFAULT_VALUE_EVICTOR_THREAD_RUN_INTERVAL_MILLIS)
+
+      // NOTE: Below lines define the behavior, so do not modify unless you know what you are
+      // doing, and update the class doc accordingly if necessary when you modify.
+
+      // 1. Set min idle objects per key to 0 to avoid creating unnecessary object.
+      // 2. Set max idle objects per key to 3 but set total objects per key to infinite
+      // which ensures borrowing per key is not restricted.
+      // 3. Set max total objects to infinite which ensures all objects are managed in this pool.
+      setMinIdlePerKey(0)
+      setMaxIdlePerKey(3)
+      setMaxTotalPerKey(-1)
+      setMaxTotal(-1)
+
+      // Set minimum evictable idle time which will be referred from evictor thread
+      setMinEvictableIdleTimeMillis(minEvictableIdleTimeMillis)
+      setSoftMinEvictableIdleTimeMillis(-1)
+
+      // evictor thread will run test with ten idle objects
+      setTimeBetweenEvictionRunsMillis(evictorThreadRunIntervalMillis)
+      setNumTestsPerEvictionRun(10)
+      setEvictionPolicy(new DefaultEvictionPolicy[InternalKafkaConsumer]())
+
+      // Immediately fail on exhausted pool while borrowing
+      setBlockWhenExhausted(false)
+
+      setJmxEnabled(jmxEnabled)
+      setJmxNamePrefix("kafka010-cached-simple-kafka-consumer-pool")
+    }
+  }
+
+  object PoolConfig {
+    val CONFIG_NAME_PREFIX = "spark.sql.kafkaConsumerCache."
+    val CONFIG_NAME_CAPACITY = CONFIG_NAME_PREFIX + "capacity"
```

Comment:
I guess I was following the style of existing code, but since we have moved to config constants in efforts on Spark 3.0, we could apply it. Will make a change.